js/src/logger.ts

```diff
@@ -675,6 +677,108 @@ export interface LogOptions<IsAsyncFlush> {
 
 export type PromiseUnless<B, R> = B extends true ? R : Promise<Awaited<R>>;
 
+interface AttachmentUploadResult {
+  signedUrl?: string;
+  metadataResponse?: Response;
+  objectStoreResponse?: Response;
+}
+
+/**
+ * Represents an attachment to be uploaded and the associated metadata. Attachment objects can be inserted anywhere in an event or feedback, allowing you to log arbitrary large data. The SDK will asynchronously upload the file to object storage and replace the Attachment object with an AttachmentReference.
+ */
+export class Attachment {
+  readonly reference: AttachmentReference;
+  /**
+   * On first access, (1) reads the attachment from disk if needed, (2) authenticates with the data plane to request a signed URL, and (3) uploads to object store.
+   */
+  readonly result: LazyValue<AttachmentUploadResult>;
+
+  private readonly data: LazyValue<Blob>;
+
+  /**
+   * Construct an attachment.
+   * @param data A string representing the path of the file on disk, or a `Blob`/`ArrayBuffer` with the file's contents. The caller is responsible for ensuring the file/blob/buffer is not modified until upload is complete.
+   * @param filename The desired name of the file in Braintrust after uploading. This parameter is for visualization purposes only and has no effect on attachment storage.
+   * @param contentType The MIME type of the file.
+   */
+  constructor(
+    data: string | Blob | ArrayBuffer,
+    filename: string,
+    contentType: string,
+  ) {
+    if (!iso.randomUUID) {
+      throw new Error("This platform does not support randomUUID");
+    }
+    const uuid = iso.randomUUID();
+    this.reference = {
+      type: BRAINTRUST_ATTACHMENT,
+      filename,
+      content_type: contentType,
+      key: `attachments/${uuid.slice(0, 2)}/${uuid}`,
+      upload_status: "uploading",
+    };
+
+    this.data = this.initData(data);
+    this.result = this.initUploader();
+  }
+
+  toJSON(): AttachmentReference {
+    return this.reference;
+  }
+
+  private initUploader(): LazyValue<AttachmentUploadResult> {
+    return new LazyValue(async () => {
+      const ret: AttachmentUploadResult = {};
+
+      try {
+        await _globalState.login({});
+        const conn = _globalState.apiConn();
+
+        // TODO error handling.
+        const [metadataResponse, data] = await Promise.all([
+          conn.post(`/attachment?${new URLSearchParams(this.reference)}`),
+          this.data.get(),
+        ]);
+        ret.metadataResponse = metadataResponse;
+        ret.signedUrl = z
+          .object({ signedUrl: z.string().url() })
+          .parse(await metadataResponse.json()).signedUrl;
+
+        // TODO multipart upload.
+        ret.objectStoreResponse = await fetch(ret.signedUrl, {
+          method: "PUT",
+          headers: {
+            "Content-Type": this.reference.content_type,
+          },
+          body: data,
+        });
+
+        if (ret.objectStoreResponse.ok) {
+          this.reference.upload_status = "done";
```

Comment:
Should we catch any errors and set the upload status to `error`? I suppose ideally we could log the error message too.

Reply:
Done